File: black-widow.py
#!/usr/bin/env python3
# coding=utf-8

"""
****************************   🕷️  Black-Widow  🕷️   ****************************
*                                                                               *
* black-widow.py -- Main black-widow executable.                                *
*                                                                               *
********************** IMPORTANT BLACK-WIDOW LICENSE TERMS **********************
*                                                                               *
* This file is part of black-widow.                                             *
*                                                                               *
* black-widow is free software: you can redistribute it and/or modify           *
* it under the terms of the GNU General Public License as published by          *
* the Free Software Foundation, either version 3 of the License, or             *
* (at your option) any later version.                                           *
*                                                                               *
* black-widow is distributed in the hope that it will be useful,                *
* but WITHOUT ANY WARRANTY; without even the implied warranty of                *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                 *
* GNU General Public License for more details.                                  *
*                                                                               *
* You should have received a copy of the GNU General Public License             *
* along with black-widow.  If not, see <http://www.gnu.org/licenses/>.          *
*                                                                               *
*********************************************************************************
"""

import app
import os
import sys

from app_plugins import args


class AppType:
    CMD = 'CMD'
    GUI = 'GUI'


# Startup
def init(app_type):
    app.utils.helpers.logger.Log.info(app.env.APP_NAME + ' ' + str(app_type) + ' started, PID=' + str(os.getpid()))
    app.utils.helpers.logger.Log.info('DEBUG is ' + str(app.env_local.APP_DEBUG))


# Main function for TEST app
def main_test():
    init(AppType.CMD)
    ip = app.utils.helpers.network.get_ip_address()
    print('ip: ' + str(ip))


# Main function for GUI app
def main_gui():
    init(AppType.GUI)
    app.gui.django_gui()


# Main function for command line app
def main_cmd(arguments):
    init(AppType.CMD)
    if arguments.django:
        django_args = str.split(arguments.django)
        app.utils.helpers.logger.Log.info('Django manager executed')
        app.utils.helpers.logger.Log.info('Django arguments: ' + str(django_args))
        app.gui.django_cmd(django_args)
        sys.exit(0)

    elif arguments.pcap:
        if arguments.pcap_int is None:
            print('Please, specify an interface! (eg. --pcap-int=wlan0)\n')
            sys.exit(1)
        app.utils.sniffing.sniff_pcap(src_file=arguments.pcap_src, interface=arguments.pcap_int,
                                      dest_file=arguments.pcap_dest, filters=arguments.pcap_filters,
                                      limit_length=arguments.pcap_limit)
    elif arguments.sql:
        if arguments.sql_url is None:
            print('Please, specify an url! (eg. --sql-url=https://black-widow.io)\n')
            sys.exit(1)
        if arguments.sql_deep:
            app.utils.sql.deep_inject_form(arguments.sql_url, arguments.sql_depth)
        else:
            app.utils.sql.inject_form(arguments.sql_url)


# Main function generic app
def main():
    if not app.utils.helpers.util.is_root():
        print("Root privileges required to run " + app.env.APP_PROC + "!\n")
        sys.exit(50)
    arguments = args.get_arguments()
    if arguments.gui:
        main_gui()
    elif arguments.test:
        main_test()
    else:
        main_cmd(arguments)


if __name__ == "__main__":
    main()

File: app/gui/web/black_widow/static/js/black-widow.js
/********************************************************************************
 *                                                                               *
 * black-widow.js -- Main black-widow javascript                                 *
 *                                                                               *
 ********************** IMPORTANT BLACK-WIDOW LICENSE TERMS **********************
 *                                                                               *
 * This file is part of black-widow.                                             *
 *                                                                               *
 * black-widow is free software: you can redistribute it and/or modify           *
 * it under the terms of the GNU General Public License as published by          *
 * the Free Software Foundation, either version 3 of the License, or             *
 * (at your option) any later version.                                           *
 *                                                                               *
 * black-widow is distributed in the hope that it will be useful,                *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                 *
 * GNU General Public License for more details.                                  *
 *                                                                               *
 * You should have received a copy of the GNU General Public License             *
 * along with black-widow.  If not, see <http://www.gnu.org/licenses/>.          *
 *                                                                               *
 ********************************************************************************/

$(function() {
    $('button.upload').click(function() {
        const $label = $(this).parent();
        $label.click();
    });
    $('input.submit').change(function() {
        const $this = $(this);
        const $form = $this.closest('form');
        if ($this.hasClass('force')) {
            $form.find('input:required').prop('required', null);
        }
        $form.submit();
    });
    $('form').submit(function() {
        if (!this.checkValidity()) {
            alert('Fill the required fields!');
            return false;
        }
        const $this = $(this);
        const $checkboxRequired = $this.find('.checkbox-required');
        if ($checkboxRequired.length > 0) {
            const $checkboxes = $checkboxRequired.find('input[type="checkbox"]:checked');
            if ($checkboxes.length === 0) {
                alert('Chose at least one ' + $checkboxRequired.attr('label') + ' !');
                return false;
            }
        }
        return true;
    });
    $('.link').click(function() {
        const $this = $(this);
        const href = $this.attr('href');
        if (href !== null && href !== undefined) {
            window.location.href = href;
        }
    })
});

const urlParams = new URLSearchParams(window.location.search);

/**
 * @param data Anything
 * @param textStatus String
 * @param jqXHR jqXHR
 * @param callback function|null
*/
const responseSuccess = function(data, textStatus, jqXHR, callback=null) {
    // console.info("[RESPONSE SUCCESS]");
    // console.log('data:');
    // console.log(data);
    // console.log('textStatus: ' + textStatus);
    if (callback !== null) {
        callback(data, textStatus, jqXHR);
    }
};

/**
 * @param jqXHR jqXHR
 * @param textStatus String
 * @param errorThrown String
 * @param callback function|null
*/
const responseError = function(jqXHR, textStatus, errorThrown, callback=null) {
    // console.error("[RESPONSE ERROR]");
    // console.log('jqXHR:');
    // console.log(jqXHR);
    // console.log('textStatus: ' + textStatus);
    // console.log('errorThrown: ' + errorThrown);
    if (callback !== null) {
        callback(jqXHR, textStatus, errorThrown);
    }
};

const request = function(
    method,
    url,
    data=null,
    successCallback=null,
    errorCallback=null
) {
    if (method === 'GET' && data !== null) {
        url += '?' + $.param(data);
        data = null;
    }
    $.ajax({
        method: method,
        url: url,
        data: data,
        headers: {
            'X-CSRFToken': $("[name=csrfmiddlewaretoken]").val()
        },
        success: function(data, textStatus, jqXHR) {
            responseSuccess(data, textStatus, jqXHR, successCallback);
        },
        error: function(jqXHR, textStatus, errorThrown) {
            responseError(jqXHR, textStatus, errorThrown, errorCallback);
        }
    });
};

/**
 *
 * @param values The row values
 * @param action The function to call on row click
 */
$.fn.insertTableRow = function (values, action = null) {
    const $table = $(this);
    let tableRow = '<tr>';
    values.forEach(value => {
        let title = value.title === undefined ? '' : 'title="' + value.title + '"';
        tableRow += '<td '+title+'>' + value.name + '</td>';
    });
    tableRow += '</tr>';
    const $tableRow = $(tableRow).appendTo($table.find('tbody:last-child'));
    if (action !== null && action !== undefined) {
        $tableRow.click(action);
    }
};

$.fn.visible = function() {
    return this.css('visibility', 'visible');
};

$.fn.invisible = function() {
    return this.css('visibility', 'hidden');
};

/**
 * Move the spinner inside the jQuery Element
 * @param message The message to show (default: "Loading...")
 */
$.fn.spinner = function (message='Loading...') {
    moveSpinner($(this), message);
    startSpinner();
};

/**
 * Move the spinner inside $parent
 * @param $parent A jQuery element
 * @param message The message to show (default: "Loading...")
 */
const moveSpinner = function($parent, message='Loading...') {
    $parent.append($('#spinner'));
    $('.spinner-text').html(message);
};

/**
 * Show the spinner
 */
const startSpinner = function() {
    $('#spinner').show();
};


/**
 * Show the spinner
 */
const showingSpinner = function() {
    return $('#spinner').is(':visible');
};

/**
 * Hide the spinner
 */
const stopSpinner = function() {
    $('#spinner').hide();
};

/**
 *
 * @param msg
 * @param type success|warning|danger
 * @param icon https://material.io/resources/icons
 * @param from
 * @param align
 */
const notify = function(msg, type='warning', icon='warning', from='top', align='right') {
    $.notify({
        icon: icon,
        message: msg
    },{
        type: type,
        timer: 3000,
        placement: {
            from: from,
            align: align
        }
    });
};

const initAccordions = function() {
    const $toggle = $('.toggle');
    $toggle.each(function() {
        const $this = $(this);
        $this.unbind();
        $this.click(function() {
            if ($this.next().hasClass('show')) {
                // Hide
                $this.find('.arrow').html('keyboard_arrow_right');
                $this.next().removeClass('show');
                $this.next().slideUp(350);
            } else {
                // Show
                const $parent = $this.parent().parent();
                $parent.find('.arrow').html('keyboard_arrow_right');
                const $innerParent = $parent.find('li .inner');
                $innerParent.removeClass('show');
                $innerParent.slideUp(350);
                $this.next().toggleClass('show');
                $this.find('.arrow').html('keyboard_arrow_down');
                $this.next().slideToggle(350);
                let $container = $this.closest('.modal');
                if ($container.length === 0) {
                    $container = $('html');
                }
                setTimeout(function () {
                    const top = $this.offset().top - $container.offset().top + $container.scrollTop();
                    $container.animate({ scrollTop: top }, 300);
                }, 350);
            }
        });
    });

};
initAccordions();
